pkg/controller/master/virtualmachine/vm_controller.go

```diff
@@ -142,6 +146,55 @@ func (h *VMController) SetOwnerOfPVCs(_ string, vm *kubevirtv1.VirtualMachine) (
 	return vm, nil
 }
 
+// SyncLabelsToVmi synchronizes the labels in the VM spec to the existing VMI without re-deployment
+func (h *VMController) SyncLabelsToVmi(_ string, vm *kubevirtv1.VirtualMachine) (*kubevirtv1.VirtualMachine, error) {
+	if vm == nil || vm.DeletionTimestamp != nil || vm.Spec.Template == nil {
+		return vm, nil
+	}
+
+	vmi, err := h.vmiCache.Get(vm.Namespace, vm.Name)
+	if err != nil && apierrors.IsNotFound(err) {
+		return vm, nil
+	} else if err != nil {
+		return nil, fmt.Errorf("get vmi %s/%s failed, error: %w", vm.Namespace, vm.Name, err)
+	}
+
+	if err := h.syncLabels(vm, vmi); err != nil {
+		return nil, err
+	}
+
+	return vm, nil
+}
+
+func (h *VMController) syncLabels(vm *kubevirtv1.VirtualMachine, vmi *kubevirtv1.VirtualMachineInstance) error {
+	vmiCopy := vmi.DeepCopy()
+	// Modification of the following reserved kubevirt.io/ labels on a VMI object is prohibited by the admission webhook
+	// "virtualmachineinstances-update-validator.kubevirt.io", so ignore those labels during synchronization.
+	// Add or update the labels of VMI
+	for k, v := range vm.Spec.Template.ObjectMeta.Labels {
+		if !strings.HasPrefix(k, kubevirt.GroupName) && vmi.Labels[k] != v {
+			if len(vmiCopy.Labels) == 0 {
+				vmiCopy.Labels = make(map[string]string)
+			}
+			vmiCopy.Labels[k] = v
+		}
+	}
+	// delete the labels existing in the vmi.Labels but not in the vm.spec.template.objectMeta.Labels
+	for k := range vmi.Labels {
+		if _, ok := vm.Spec.Template.ObjectMeta.Labels[k]; !strings.HasPrefix(k, kubevirt.GroupName) && !ok {
+			delete(vmiCopy.Labels, k)
+		}
```

Comment:
Does removing the kubevirt-specific labels will cause any side effect?

Reply:
Kubevirt has a webhook to avoid those labels from deleting.